src/node/mini_miner.cpp

```diff
@@ -0,0 +1,265 @@
+
+#include <node/mini_miner.h>
+
+#include <consensus/amount.h>
+#include <policy/feerate.h>
+#include <primitives/transaction.h>
+#include <timedata.h>
+#include <util/check.h>
+#include <util/moneystr.h>
+
+#include <algorithm>
+#include <numeric>
+#include <utility>
+
+namespace node {
+
+MiniMiner::MiniMiner(const CTxMemPool& mempool, const std::vector<COutPoint>& outpoints)
+{
+    LOCK(mempool.cs);
+    requested_outpoints = outpoints;
+    // Find which outpoints to calculate bump fees for.
+    // Anything that's spent by the mempool is to-be-replaced
+    // Anything otherwise unavailable just has a bump fee of 0
+    for (const auto& outpoint : outpoints) {
+        if (const auto ptx{mempool.GetConflictTx(outpoint)}) {
+            // The conflicting transaction is to-be-replaced
+            to_be_replaced.insert(ptx->GetHash());
+        }
+
+        if (!mempool.exists(GenTxid::Txid(outpoint.hash))) {
+            // This UTXO is either confirmed or not yet submitted to mempool.
+            // In the former case, no bump fee is required.
+            // In the latter case, we have no information, so just return 0.
+            this->bump_fees.emplace(std::make_pair(outpoint, 0));
+        } else {
+            // This UTXO is unconfirmed, in the mempool, and available to spend.
+            auto it = outpoints_needed_by_txid.find(outpoint.hash);
+            if (it != outpoints_needed_by_txid.end()) {
+                it->second.push_back(outpoint);
+            } else {
+                std::vector<COutPoint> outpoints_of_tx({outpoint});
+                outpoints_needed_by_txid.emplace(std::make_pair(outpoint.hash, outpoints_of_tx));
+                // Instead of operating on the entire mempool, just run the mining algorithm on the
+                // cluster of relevant transactions, which we'll store in mapModifiedTx.
+            }
+        }
+    }
+    // Calculate the cluster and construct the entry map.
+    std::vector<uint256> txids_needed;
+    std::transform(outpoints_needed_by_txid.cbegin(),
+                   outpoints_needed_by_txid.cend(),
+                   std::back_inserter(txids_needed),
+                   [](const auto& pair) { return pair.first; });
+    const auto& cluster = mempool.CalculateCluster(txids_needed);
+    for (const auto& txiter : cluster) {
+        if (to_be_replaced.find(txiter->GetTx().GetHash()) == to_be_replaced.end()) {
+            auto [mapiter, success] = entries_by_txid.emplace(std::make_pair(txiter->GetTx().GetHash(), MockMempoolEntry(txiter)));
+            assert(success);
+            entries.push_back(mapiter);
+        } else {
+            auto outpoints_it = outpoints_needed_by_txid.find(txiter->GetTx().GetHash());
+            if (outpoints_it != outpoints_needed_by_txid.end()) {
+                for (const auto& outpoint : outpoints_it->second) {
+                    this->bump_fees.emplace(std::make_pair(outpoint, 0));
+                }
```

Comment:
```suggestion
                for (const auto& outpoint : outpoints_it->second) {
                    bump_fees.emplace(outpoint, 0);
                }
                outpoints_needed_by_txid.erase(outpoints_it);
```
I'm not sure about this, but doing the `erase` here would be consistent with the `!mempool.exists()` case above (an entry for this transaction, none of whose outpoints we will use) is _not_ added to `outpoints_needed_by_txid`).